### ISSUE TITLE
Hero's path not converted

### ISSUE THREAD

**Issue by nedex** [WemI0/BOTW_SaveConv#1]
The [Hero's Path](https://zelda.gamepedia.com/Hero%27s_Path_Mode) informations (`tracker` folder) doesn't get converted properly.
I also tried to leave the files unconverted without success.
Any way to decrypt sav files to help with development?

**Comment by WemI0** [comment]
Can you sent a savegame with the "Hero's Path" informations ?

**Comment by nedex** [comment]
You can easily get any savegame from the internet
[Wii U](https://drive.google.com/file/d/0B5p7e1g-NHIvSXdQekktWGFBaVU/view) (I couldn't find any Wii U savegame with 32 blocks - eg. - longer than 200 hours)
[Switch](https://gbatemp.net/threads/nintendo-switch-save-file-megatutorial.503831/#post-7981345)

These are the informations I have:
- Files are named `trackblock00.sav` to `trackblock31.sav` (`trackblock_hardXX.sav` for hard mode) in the tracker folder
- First 4 bytes must not be modified (0x0 to 0x3)
- The two following bytes (0x4 and 0x5) must be switched (for complete trackblock file: `012C` on Wii U and `2C01` on Switch)
- The conversion can continue as always from 0x8
- No need to search for text

I'll keep you updated if I have more informations.

**Comment by suzsuzuki** [comment]
Any news about the Hero's Path?

**Comment by stratusfear** [comment]
Hero's Path data would show up for me until the first time I saved after importing my converted save file. After loading the game again, all the path data would be gone. Following @nedex instructions has fixed it for me.

**Comment by victor2203** [comment]
Hello.
I have been trying to modify the trackblockXX.save files but I don't know how to do it with a hex editor. The conversion from Wii U to Switch format was fine, except the hero's path. It doesn't show anything of my hours and hours WiiU savegame. Could somebody help me is I send my trackblockXX.save files please? Thanks.

**Comment by shengolol** [comment]
I can't get it to work with @nedex instructions. My savegame has files till block 14. Block 14 does not even have "012C" at all. I would be very thankful for clear instructions.

**Comment by nedex** [comment]
> Block 14 does not even have "012C" at all. I would be very thankful for clear instructions.
Because it's the last one, it isn't a "complete block".

The code should be modified since this instructions aren't for all end users...

**Comment by shengolol** [comment]
I understood the instructions concerning 0x0 to 0x7 now, I still have some questions though:
1) Should the complete block have "00" or "01" at 0x8 for the Switch? I am missing a reference save file for the switch, and the 100% files I have has "01".
2) How should the last incomplete block be converted from Wii U to Switch?

**Comment by rossome89** [comment]
I'm trying to sort this out just now too. The headers of my "last" block (08) is completely different from 00 to 07.

**Comment by rossome89** [comment]
@shengolol 

> 1. Should the complete block have "00" or "01" at 0x8 for the Switch? I am missing a reference save file for the switch, and the 100% file I have has "01".

I've been collecting a number of real Switch saves to compare. A number of the earlier blocks sometimes start off writing 01 at 0x8 but then after a period of time it changes to 11. This could be referencing a newer version of BoTW?

Here are a couple of "last incomplete block"s I have found/been given from real genuine Switch saves:

Switch 1:
`
0d | 0d | 0d | 00 | ef | 00 | 00 | 00 | 11 | 00 | 00 | 00 | 00 | 00 | 00 | 00
`

Switch 2:
`
0b | 0b | 2d | 00 | 90 | 00 | 00 | 00 | 01 | 00 | 00 | 00 | 00 | 00 | 00 | 00
`

0x4 is the value that's throwing me off. I can't work out what it's for and it seems to vary wildly from file to file. I thought it might be "Next line to write values at" but in one of my examples (Switch 1) the value at 0x4 is much bigger than the same value in Switch 2, even though the TrackerBlock in Switch 2 has more completed lines in it.

**Comment by shengolol** [comment]
@rossome89 
Thank you for this information. Could you add which trackblock file the headers came from? (trackblockX). Better yet, uploading the saves could maybe help in the future, too!
I will dump the save from my brother's switch so we will have fourth save.

**Comment by nedex** [comment]
Bytes 0x4 and 0x5 contains the length of the save, where 012C is the max.
**Just need to swap them** between Wii U and Switch, **and do the "standard" the conversion** from 0x08.
I don't really think there's need for further investigation if you just need to do the conversion.

**Comment by shengolol** [comment]
> 
> 
> Bytes 0x4 and 0x5 contains the length of the save, where 012C is the max.
> **Just need to swap them** between Wii U and Switch, **and do the "standard" the conversion** from 0x08.
> I don't really think there's need for further investigation if you just need to do the conversion.

Unfortunately it does not simply not work. If it were so easy

**Comment by rossome89** [comment]
It doesn't work for me either. It's like the data is all wrong and placing
the points at locations that are out of bounds.

@shengolol I will try and email the files to the email address on your
account soon.

**Comment by nedex** [comment]
> Unfortunately it does simply not work. If it were so easy
Well, it worked for me after several tests and then for somebody else so we were lucky.

**Comment by rossome89** [comment]
I think I've figured it out. Bare with me as I've not used hex extensively before. The software goes through each address location and swaps the value at `0x0 `with the value at `0x3` and `0x1 with 0x2`, `0x4 with 0x7` and `0x5 with 0x6`, `0x8 with 0xB` and `0x9 with 0xA`, and finally `0xC with 0xF` and `0xD with 0xE`.

That works great, but it also does it to the headers, which screws them up and breaks them (by placing the values that should be at `0x4` and `0x5` at `0x6` and `0x7` and so on).

The co-ordinate details from address 50 onwards DO need to be flipped (Usually the first coordinate at address 50 across `0x1 0x2 0x3 0x4` on every new save file is `7D 87 8D 84` on the Switch, as opposed to `84 8D 87 7D` on the WiiU).

So I think (and I haven't had a chance to test this yet) that if we use the .exe to convert the BOTW save, and then manually edit the headers for each trackblockXX.sav file (swapping the WiiU headers back in and then manually changing 0x4 and 0x5 to `2C` and `01`, and so on), we should end up with the correct headers and also valid location information.

I'll report back soon.

**Comment by nedex** [comment]
Yes, that's why i wrote that you should modify the program for trackblocks to start only from 0x08 and manually edit the first 8 bytes...

**Comment by rossome89** [comment]
@nedex I think when some of us followed the instructions we were left with no data at 0x4 or 0x5 so couldn't swap them.

**Comment by shad00m** [comment]
Is any method working so far?
I just noticed I can't unlock all the memories.

**Comment by nedex** [comment]
@shad00m this is for the hero's path. Your problem seems to be related to the main conversion.

**Comment by xombiemp** [comment]
I can confirm that @nedex 's solution works.

I used DeltaJordan's fork, https://github.com/DeltaJordan/BotW-Save-Manager, and after I ran the conversion tool, I used a hex editor on all the trackblockXX.sav files and I copied bytes 0x0 to 0x7 from the original Wii U save and pasted it over the converted Switch save in the same place. Then I swapped values 0x4 and 0x5. (Don't get caught up on the values of 0x4 and 0x5, just swap them) After doing this with all the files and restoring that save, I can now see my full hero's path on the map.
